### PR TITLE
Sandbox/srivas/caliper

### PR DESCRIPTION
**Pull request by sergiorg-hpc**
**Description**

The commit fixes an issue to compile with the latest versions of Caliper and adds support for report flushing instrumentation.

- [x] Update CMakeLists.txt file to remove unnecessary link flags.
- [x] Add instrumentation for report flushing.

**How to test this?**

Testing can be accomplished, for instance, with Neurodamus-Neocortex (note the `+caliper` for CoreNEURON):

```bash
spack install -v neurodamus-neocortex@develop+coreneuron ^coreneuron@1.0c+caliper ^neuron@develop
module purge && module load unstable neurodamus-neocortex/develop py-neurodamus
...
CALI_CONFIG=runtime-report,profile.mpi srun dplace special -mpi -python $NEURODAMUS_PYTHON/init.py
```

**Test System**
 - OS: Red Hat Enterprise Linux Server v7.6
 - Compiler: GCC 9.3.0 + HPE-MPI v2.22 HMPT
 - Version: master branch
 - Backend: Phase II nodes

**Dependencies**
This feature depends on a Spack branch available here:
https://github.com/BlueBrain/spack/pull/1156

CI_BRANCHES:NEURON_BRANCH=master,

### CODE REVIEW

**Comment by sergiorg-hpc** [comment]
**Important note:** This is still work-in-progress.

**Review by alkino — COMMENTED** [review]
No thoughts about it, I don't know caliper.

File: coreneuron/CMakeLists.txt [hunk 1]
```diff
@@ -181,7 +181,6 @@ target_link_libraries(
   ${link_cudacoreneuron}
   ${CUDA_LIBRARIES}
   ${CALIPER_LIB}
-  ${CALIPER_MPI_LIB}
```
[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply coreneuron/CMakeLists.txt:184]: This is still work-in-progress! Anyhow, @iomaganaris guided me a bit (thanks a lot) and we checked that Caliper contains MPI wrappers et al.:

```Bash
nm /gpfs/bbp.cscs.ch/data/project/proj16/srivas/spack/software/linux-rhel7-x86_64/gcc-9.3.0/caliper-2.4.0-7xw5pm/lib64/libcaliper.so | grep mpi
0000000000085550 T cali_mpi_init
00000000000822e0 t _GLOBAL__sub_I_aggregate_over_mpi.cpp
00000000000822b0 t _GLOBAL__sub_I_machine_mpi.cpp
...
000000000041f968 b _ZZN12_GLOBAL__N_110push_mpifnEPN4cali7CaliperEbPKcE7call_id
0000000000419960 b _ZZN4cali12mpiwrap_initEPNS_7CaliperEPNS_7ChannelEE10s_bindings
0000000000419948 b _ZZN4cali12mpiwrap_initEPNS_7CaliperEPNS_7ChannelEE13s_binding_pos
00000000000b4670 t _ZZN4cali12mpiwrap_initEPNS_7CaliperEPNS_7ChannelEENKUlS1_S3_E_clES1_S3_.isra.0
0000000000074966 t _ZZN4cali12mpiwrap_initEPNS_7CaliperEPNS_7ChannelEENKUlS1_S3_E_clES1_S3_.isra.0.cold
```
Nonetheless, I still need to double-check whether the information provided makes sense or not.

**Comment by sergiorg-hpc** [comment]
Just a quick update about whether MPI profiling is available. It seems that the profiling works, as illustrated below:

```
$ CALI_CONFIG=runtime-report,profile.mpi srun -n 40 dplace special-core --read-config /.../output_caliper/sim.conf --mpi
...
[REPORTS] [info] :: Initializing PARALLEL implementation...
[REPORTS] [info] :: Initializing communicators and preparing SONATA datasets
...
Path                           Min time/rank Max time/rank Avg time/rank Time %    
main                                0.232202      0.259930      0.240925  1.444035 
  MPI_Finalize                      0.000009      0.000015      0.000012  0.000073 
  MPI_File_close                    0.009219      0.009229      0.009223  0.055280 
  checkpoint                        0.000004      0.000006      0.000005  0.000027 
  output-spike                      0.001963      0.005134      0.002335  0.013994 
    MPI_File_write_at_all           0.000221      0.000227      0.000223  0.001337 
    MPI_Exscan                      0.000211      0.000214      0.000213  0.001274 
    ...
  simulation                        0.007590      0.008417      0.007820  0.046869 
    MPI_Barrier                     0.000003      0.001157      0.001097  0.006573 
    flush_reports                   0.028947      0.034984      0.030932  0.185399 
      MPI_Type_size_x               0.000039      0.000055      0.000047  0.000284 
      MPI_Get_elements_x            0.000115      0.000137      0.000126  0.000757 
      MPI_File_write_at_all         0.389533      0.504861      0.434999  2.607253 
      ...
    spike-exchange                  0.002151      0.003928      0.002275  0.013637 
      communication                 0.002195      0.002417      0.002302  0.013800 
        MPI_Allgather               0.368152      0.483335      0.424084  2.541832 
  ...
  load-model                        3.803580      8.402059      6.340712 38.004347 
    MPI_Allreduce                   0.013192      4.612440      2.074944 12.436596 
  MPI_Allreduce                     0.001537      0.006939      0.004167  0.024976
  ```
  
On the other hand, running without the `profile.mpi` configuration, the result is different, as expected:

```
$ CALI_CONFIG=runtime-report srun -n 40 dplace special-core --read-config /.../output_caliper/sim.conf --mpi
...
[REPORTS] [info] :: Initializing PARALLEL implementation...
[REPORTS] [info] :: Initializing communicators and preparing SONATA datasets
...
Path                     Min time/rank Max time/rank Avg time/rank Time %    
main                          4.075285      4.084183      4.077681 23.069888 
  checkpoint                  0.000003      0.000005      0.000004  0.000024 
  output-spike                0.259590      0.259612      0.259603  1.468732 
  simulation                  0.007573      0.012089      0.008092  0.045782 
    flush_reports             1.129518      1.147010      1.134374  6.417834 
    spike-exchange            0.002508      0.007206      0.002969  0.016797 
      communication           0.356627      0.512863      0.442750  2.504903 
      imbalance               0.758353      0.954505      0.853334  4.827823 
    timestep                  0.116003      0.136037      0.125003  0.707215 
      deliver_events          0.640983      0.687554      0.661331  3.741546 
  finitialize                 0.644216      0.815266      0.684544  3.872878 
    spike-exchange            0.000002      0.000015      0.000006  0.000034 
      communication           0.000196      0.000207      0.000204  0.001152 
      imbalance               0.000056      0.170726      0.130704  0.739470 
    cur-StochKv               0.000281      0.000627      0.000415  0.002349 
    ...
  load-model                  9.241541      9.250388      9.248515 52.324402 
```

File: coreneuron/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (iomaganaris) [reply coreneuron/CMakeLists.txt:184]: I remember in the past I had the same issues as Sergio with `Caliper`. Basically after this commit https://github.com/LLNL/Caliper/commit/84d06546b7beeda831c54f50cbe0d6b882fb00fa#diff-d1da5e5cf11e9574264239e71ba6a518e9fba2e1e065c43d4a963eef6d1837a3 they removed the `caliper-mpi` libraries, so everything is bundled in `libcaliper.so` as far as I remember and tested

[the remark this answers is not in the document]
Comment (sergiorg-hpc) [reply coreneuron/CMakeLists.txt:184]: Thanks a lot, @iomaganaris!

**Comment by pramodk** [comment]
@sergiorg-hpc : you converted this to draft. Isn't this ready for merge?

**Comment by sergiorg-hpc** [comment]
> @sergiorg-hpc : you converted this to draft. Isn't this ready for merge?

I think so too, but I'm still in the process of running and collecting results with Caliper outputs, so I wanted to be cautious in case we are missing something.

If you consider that having the instrumentation during flush is enough, we can go ahead and merge this, for sure!

**Pull request merged**